### DISCUSSION TITLE
Changing password when ERRCONNECT_PASSWORD_MUST_CHANGE

### DISCUSSION THREAD

**Discussion by ecos-ps** [FreeRDP/FreeRDP#7136] [Q&A] [answered]
Hi there,

We are facing the problem, that a user wants to connect to a Microsoft RDS where he has to change his password but only gets the error "ERRCONNECT_PASSWORD_MUST_CHANGE" instead of a possibility to change his password.
Is there any way to prevent this and show password edit dialog?

### Reproduction

1. Create test user "toni.tester" on a Microsoft RDS
2. Go to settings of user
3. Click "User has to change password at next login" (Sorry, i don't know the right tab and translation since we are working on a German environment) Attribute msDS-User-Account-Control-Computed is now set to 8388608 (or PASSWORD_EXPIRED)
5. Try to login with "toni.tester" using freerdp

### Log
I have XXXed out our internal domain and server credentials.
```
$ /usr/bin/xfreerdp /cert-ignore /bpp:'24' /kbd:'0x407' /t:'RDP Test' +compression /d:'XXXXXXX' +decorations /floatbar:sticky:off,default:hidden,show:fullscreen +fonts /network:lan /log-level:ERROR /f /multimon /rfx /rfx-mode:video /sound:sys:pulse,format:6 /microphone:sys:pulse,format:6 /v:'XXXXXXXX'
Username: toni.tester
Password:
[13:49:01:723] [9067:9068] [ERROR][com.freerdp.core] - nla_recv_pdu:freerdp_set_last_error_ex ERRCONNECT_PASSWORD_MUST_CHANGE [0x00020013]
[13:49:01:723] [9067:9068] [ERROR][com.freerdp.core.rdp] - rdp_recv_callback: CONNECTION_STATE_NLA - nla_recv_pdu() fail
[13:49:01:723] [9067:9068] [ERROR][com.freerdp.core.transport] - transport_check_fds: transport->ReceiveCallback() - -1
```

Since there are errors concerning NLA as well, we have already tried to disable or enable other NLA options, but it didn't work.

Can someone pinpoint us to the right direction?

Thanks and have a nice and happy coding week!

**Comment by akallabeth** [comment, marked as answer]
you might try with `-sec-nla` and connect with `TLS` (if the server allows it) but `FreeRDP` currently does not handle that case.

**Reply by ecos-ps** [reply]
`-sec-nla` lets us connect to our server as well as `-sec-tls` in our test environment, but we are not allowed to connect to the server without NLA or TLS in production, so this won't help atm.
I will write a feature request.
Thanks for your fast reply and help!

**Reply by ecos-ps** [reply]
So i nearly send a feature request to you, but considered giving it a try looking for another workaround. I haven't thought about searching in German by now, but found a workaround and i think it is clever to post it here, even when it is in German, so that other users will find it as well.
I try to translate it in short terms:
[RDP Change Password through RDWeb](https://www.windowspro.de/wolfgang-sommergut/passwort-aendern-ueber-remotedesktop-rd-web-access)
It is not possible (even through MSTSC.exe) to change password when it expired. You have to call your admin **OR** your admin configured RDWeb such that every user can change his password on its own **OR** disable NLA **( NOT RECOMMENDED! )**

### Instructions for RDWeb
Enable `PasswordChangeEnabled` in IIS Manager from Default Website => RDWeb => Pages by setting value on `true`
**OR (when no IIS Manager installed)**
Change policy in file `C:\Windows\Web\RDWeb\Pages\web.config` to `<add key="PasswordChangeEnabled" value="true" />`

Now it is possible for every user to change his password through `https://<server-FQDN>/RDWeb/Pages/de-DE/password.aspx` (don't forget to change to your lanfuage, for english use en_EN or en_US)
